src/datachain/lib/signal_schema.py

```diff
@@ -749,3 +754,108 @@ def _build_tree_for_model(
+        with the original schema.
+
+        Args:
+            *columns (str): The columns to include in the partial schema.
+
+        Returns:
+            SignalSchema: The new partial schema.
+        """
+        serialized = self.serialize()
+        custom_types = serialized.get("_custom_types", {})
+
+        schema: dict[str, Any] = {}
+        schema_custom_types: dict[str, CustomType] = {}
+
+        data_model_bases: Optional[list[tuple[str, str, Optional[str]]]] = None
+
+        for column in columns:
+            parent_type_origin, parent_type_partial = "", ""
+            for i, signal in enumerate(column.split(".")):
+                if i == 0:
+                    if signal not in serialized:
+                        raise SignalSchemaError(
+                            f"Column {column} not found in the schema"
+                        )
+
+                    parent_type_origin = serialized[signal]
+                    parent_type_partial = _type_name_to_partial(parent_type_origin)
+
+                    schema[signal] = parent_type_partial
+                    continue
+
+                if parent_type_origin not in custom_types:
+                    raise SignalSchemaError(
+                        f"Custom type {parent_type_origin} not found in the schema"
+                    )
+
+                custom_type = custom_types[parent_type_origin]
+                signal_type = custom_type["fields"].get(signal)
+                if not signal_type:
+                    raise SignalSchemaError(
+                        f"Field {signal} not found in custom type {parent_type_origin}"
+                    )
+
+                partial_type = _type_name_to_partial(signal_type)
+
+                if parent_type_partial in schema_custom_types:
+                    schema_custom_types[parent_type_partial].fields[signal] = (
+                        partial_type
+                    )
+                else:
+                    if data_model_bases is None:
+                        data_model_bases = SignalSchema._get_bases(DataModel)
+
+                    partial_type_name, _ = ModelStore.parse_name_version(partial_type)
+                    schema_custom_types[parent_type_partial] = CustomType(
+                        schema_version=2,
+                        name=partial_type_name,
+                        fields={signal: partial_type},
+                        bases=[
+                            (partial_type_name, "__main__", partial_type),
+                            *data_model_bases,
+                        ],
+                    )
+
+                parent_type_origin, parent_type_partial = signal_type, partial_type
+
+        if schema_custom_types:
+            schema["_custom_types"] = {
+                type_name: ct.model_dump()
+                for type_name, ct in schema_custom_types.items()
+            }
+
+        return SignalSchema.deserialize(schema)
```

Comment:
can we construct the object itself instead of deserializing it?

Reply:
Sure we can, but I think it will be almost identical to `deserialize` 👀 IMO it is better to reuse existing function.